### PR TITLE
Fix favicon download from URL with non-standard port.

### LINKED ISSUE #5001: Download favicon.ico not possible with port number in URL
## Overview
Downloading a favicon.ico is not possible if an URL contains an port number.
Example: `https://192.168.1.1:1234/favicon.ico`

## Steps to Reproduce
[NOTE]: # ( Provide a simple set of steps to reproduce this bug. )
1. Open in browser  `https://192.168.1.1:1234/favicon.ico` (Favicon is visible, server listen to port 1234)
2. Open in browser  `https://192.168.1.1/favicon.ico` (Favicon is NOT visible, because server doesn't listen to port 443)
3. Make an entry with `https://192.168.1.1:1234/login` (example)
4. Select option 'Download Favicon' -> It shows a dialog with a status 'Download Failed'

## Expected Behavior
[NOTE]: # ( Tell us what you expected to happen )
In the last step 4 the favicon should be downloaded.

## Actual Behavior
[NOTE]: # ( Tell us what actually happens )
In the last step 4 the favicon is not downloaded, however it is accessible in a browser (Step 1)

## Context
[NOTE]: # ( Give us any additional information you may have. )
Probably the port number is not 'added' after the FQDN which is used to search the favicon.

[NOTE]: # ( Paste debug info from Help → About here )
KeePassXC - 2.6.0
Revision: REVISION

[NOTE]: # ( Pick choices based on your environment )
Operating System: Linux Mint 19.3 and macOS
Desktop Env: Cinnamon
Windowing System: X11/Wayland ???

### PR DESCRIPTION
**Pull request by libklein**
Fixes #5001 and #2843.

The favicon download URL was constructed from scheme and host only. This is fixed by simply replacing the path of the original URL with "/favicon.ico", thus keeping scheme, host, auth and port intact.

**Further modification**: URL's with a non-http schema are now rejected. (See #2843)

## Testing strategy
Tested URLs mentioned in #5001 and several well-known domains.

## Type of change
[NOTE]: # ( Please remove all lines which don't apply. )
- ✅ Bug fix (non-breaking change that fixes an issue)
- ✅ Breaking change (causes existing functionality to change)

### CODE REVIEW

**Comment by libklein** [comment]
As a side note: Perhaps a refactor of the IconDownloader is in order. Right now it tries to download the favicon not only from the host itself, but also from the second-level domain. I'd argue that it would make more sense to consider the host only, i.e. get rid of the second-level domain request. 

Moreover, it would be nice to support the W3C recommendation regarding favicons, i.e. to (rudimentarily) parse the page and extract the various <link rel="icon" /> tags (c.f. [wikipedia](https://en.wikipedia.org/wiki/Favicon)).

Any feelings towards this? I'd be happy to provide an implementation, just wanted to hear your opinion on the matter first.

**Comment by droidmonkey** [comment]
We are not interested in parsing html... but maybe if we just did a very simple regex version.

**Comment by droidmonkey** [comment]
Also when using the scheme functions calls of QUrl you do not add the ://

**Comment by libklein** [comment]
> Also when using the scheme functions calls of QUrl you do not add the ://

Sorry about that. Will fix it tomorrow (or perhaps wait for #5504, if that is preferred) and then rebase on top of that. 

> We are not interested in parsing html... but maybe if we just did a very simple regex version.

I agree, that seems like a bit much given that most cases are covered by "/favicon.ico" anyway. What about getting rid of the second-level domain request?

**Comment by droidmonkey** [comment]
If we regex the html we can get rid of the second level domains.

**Comment by libklein** [comment]
I've rebased onto #5504, fixed the schema bug and added a test case covering some corner cases, such as query strings, username/password in url and nested domains.

**Comment by libklein** [comment]
Hi, i just wanted to ask whats the state on this? Are you planning to merge this (soon) or is there something wrong with the PR?

**Comment by droidmonkey** [comment]
Its in the review queue, you don't have to do anything

File: tests/TestIconDownloader.cpp [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    downloader.setUrl(QString(url));
+    QList<QString> resolved_urls;
+    for (const auto& resolved_url : downloader.m_urlsToTry) {
+        resolved_urls.push_back(resolved_url.toString());
+    }
+    return resolved_urls;
+}
+
+void TestIconDownloader::testFaviconResolve()
+{
```
Comment (phoerious) [inline tests/TestIconDownloader.cpp:19]: You should take a look at data-driven testing in Qt.

**Comment by phoerious** [comment]
Perhaps we can even split all the downloading functionality into a separate binary for address space separation as an additional mitigation against parser exploits. Then I'd even be fine with doing some more advanced logic than regex.

**Comment by libklein** [comment]
Hi,

sorry to bother you guys again, but are there any plans to merge this in the near future?

Best regards,
Patrick.

**Comment by droidmonkey** [comment]
Yes sorry this has been on my review list. Getting through it!

File: src/gui/IconDownloader.cpp [hunk 2]
```diff
@@ -73,16 +73,24 @@ namespace
...
         return;
     }
 
     m_redirects = 0;
     m_urlsToTry.clear();
 
+    // If no scheme is specified, fall back to http as http is usually redirected to https, but not vice versa
     if (url.scheme().isEmpty()) {
-        url.setUrl(QString("https://%1").arg(url.toString()));
+        url.setScheme("http");
```
Comment (droidmonkey) [inline src/gui/IconDownloader.cpp:86]: Any specific reason for defaulting to http? Generally we default to https.

File: src/gui/IconDownloader.cpp [hunk 3]
```diff
@@ -91,16 +99,15 @@ void IconDownloader::setUrl(const QString& entryUrl)
...
+    hostIsIp =
+        std::any_of(hostAddressess.begin(), hostAddressess.end(), [&fullyQualifiedDomain](const QHostAddress& addr) {
+            return addr.toString() == fullyQualifiedDomain;
+        });
 
     // Determine the second-level domain, if available
     QString secondLevelDomain;
     if (!hostIsIp) {
-        secondLevelDomain = getSecondLevelDomain(m_url);
+        secondLevelDomain = getSecondLevelDomain(url);
```
Comment (droidmonkey) [inline src/gui/IconDownloader.cpp:110]: Does this preserve the port number? Should it?

**Review by droidmonkey — CHANGES_REQUESTED** [review]

File: src/gui/IconDownloader.cpp (hunk 2 above)
Comment (libklein) [reply src/gui/IconDownloader.cpp:86]: I have defaulted to http:// here as that maximizes the probability of finding the correct service. However i agree that it would be better to default to HTTPS, especially given that the user may specify sensitive data as part of the url. I will update the PR as soon as the other issue has been discussed.

File: src/gui/IconDownloader.cpp (hunk 3 above)
Comment (libklein) [reply src/gui/IconDownloader.cpp:110]: It preserves port number, username/password, query, ..., i.e., all non-host fields of the URL. I see the issue here, still, i think we should at least consider the URL as a candidate. Perhaps adding two URLs, one blank request to the second level domain with the original query and one that keep all original data intact would be reasonable?

Comment (droidmonkey) [reply src/gui/IconDownloader.cpp:110]: Yes I agree, that is in the spirit of maximizing success. Generally speaking this should never be hit since it is rare that a subdomain does not provide its own icon.
